fleetshard/pkg/centralreconciler/reconciler.go

```diff
@@ -10,53 +10,66 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime"
-	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
-	ctrl "sigs.k8s.io/controller-runtime"
+	"k8s.io/utils/pointer"
 	ctrlClient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sync/atomic"
 )
 
-func Synchronize(devEndpoint string, clusterID string) {
-	client, err := fleetmanager.NewClient(devEndpoint, clusterID)
-	if err != nil {
-		glog.Fatal("failed to create fleetmanager client", err)
-	}
+const (
+	FreeStatus int32 = iota
+	BlockedStatus
+)
 
-	// TODO(create-ticket): Add filter in the REST Client to only receive a specific state
-	list, err := client.GetManagedCentralList()
-	if err != nil {
-		glog.Fatalf("failed to list centrals for cluster %s: %s", clusterID, err)
-	}
+// CentralReconciler reconciles the central cluster
+type CentralReconciler struct {
+	client             ctrlClient.Client
+	central            private.ManagedDinosaur
+	fleetManagerClient *fleetmanager.Client
+	receiveCh          chan *private.ManagedDinosaur
+	errorCh            chan ReconcilerError
+	stopCh             chan struct{}
+	blocked            *int32
+}
 
-	statuses := make(map[string]private.DataPlaneDinosaurStatus)
-	for _, remoteCentral := range list.Items {
-		glog.Infof("received cluster: %s", remoteCentral.Metadata.Name)
+// TODO: Correct error creation?
+type ReconcilerError struct {
+	Central private.ManagedDinosaur
+	Err     error
+}
 
-		reconciler := NewClusterReconciler()
-		status, err := reconciler.Reconcile(context.Background(), remoteCentral)
-		if err != nil {
-			glog.Fatalf("failed to reconcile central %s: %s", remoteCentral.Metadata.Name, err)
+// TODO(create-ticket): Setup local watch on Central's kube API resources to set new updates statues?
+// TODO(create-ticket): Graceful shutdown?
+func (r CentralReconciler) Start() {
+	for {
+		select {
+		case central := <-r.receiveCh:
+			go func() {
+				//TODO: cancellation?
+				err := r.Reconcile(context.Background(), *central)
+				if err != nil {
+					r.errorCh <- ReconcilerError{Err: err}
+				}
+			}()
+		case <-r.stopCh:
+			return
 		}
-
-		statuses[remoteCentral.Id] = *status
-	}
-
-	resp, err := client.UpdateStatus(statuses)
-	if err != nil {
-		glog.Fatal(err)
 	}
-	glog.Infof(string(resp))
 }
 
-// ClusterReconciler reconciles the central cluster
-type ClusterReconciler struct {
-	client ctrlClient.Client
+func (r CentralReconciler) ReceiveCh() chan *private.ManagedDinosaur {
+	return r.receiveCh
 }
 
-func (r ClusterReconciler) Reconcile(ctx context.Context, remoteCentral private.ManagedDinosaur) (*private.DataPlaneDinosaurStatus, error) {
+func (r CentralReconciler) Reconcile(ctx context.Context, remoteCentral private.ManagedDinosaur) error {
+	// Only allow to start reconcile function once
+	if !atomic.CompareAndSwapInt32(r.blocked, FreeStatus, BlockedStatus) {
+		return errors.New("Reconciler still busy, wait until current reconciliation has stopped.")
```

Comment:
`wait until ...` sounds like a blocking operation.

```suggestion
		return errors.New("Reconciler still busy, skipping reconciliation attempt.")
```

Reply:
done